### PR TITLE
Migrate transform-vdom to TypeScript

### PR DESCRIPTION
**Pull request by captainsafia**
Closes https://github.com/nteract/nteract/issues/3587.

I left some comments and questions inline.

### CODE REVIEW

File: packages/transform-vdom/src/object-to-react.ts [hunk 1]
```diff
@@ -35,16 +33,22 @@
...
+
+export interface VDOMEl {
+  tagName: string; // Could be an enum honestly
+  children: VDOMNode;
+  attributes: Attributes;
+  key: number | string | null;
 };
 
-type ReactArray = Array<React$Element<any> | ReactArray | string>;
+interface ReactArray extends ConcatArray<React.ReactElement<any> | string | ReactArray> {};
```
Comment (captainsafia) [inline packages/transform-vdom/src/object-to-react.ts]: Is this type supposed to be equivalent to the type defined in `React.ReactArray`?

File: packages/transform-vdom/src/object-to-react.ts [hunk 2]
```diff
@@ -95,7 +99,7 @@ export function objectToReactElement(obj: VDOMEl): React$Element<any> {
       if (args[1] === undefined) {
         args[1] = null;
       }
-      args = args.concat(arrayToReactChildren(children));
+      args = args.concat(arrayToReactChildren(children) as any);
```
Comment (captainsafia) [inline packages/transform-vdom/src/object-to-react.ts]: There were some complaints here about the recursive definition in the `ReactArray` type.

File: packages/transform-vdom/src/object-to-react.ts [hunk 3]
```diff
@@ -35,16 +33,22 @@
...
+  tagName: string; // Could be an enum honestly
+  children: VDOMNode;
+  attributes: Attributes;
+  key: number | string | null;
 };
 
-type ReactArray = Array<React$Element<any> | ReactArray | string>;
+interface ReactArray extends ConcatArray<React.ReactElement<any> | string | ReactArray> {};
+
+type VDOMNode = VDOMEl | string | any;
```
Comment (captainsafia) [inline packages/transform-vdom/src/object-to-react.ts]: I had trouble getting the circular reference to `VDOMNode` working here. TypeScript outlines a technique for using circular references in Array-like types which I used above but not for union types. I don't know much about the goal of this type to be able to define it without using circular types.

File: packages/transform-vdom/src/object-to-react.ts (hunk 1 above)
Comment (rgbkrk) [reply packages/transform-vdom/src/object-to-react.ts]: I think so.

**Comment by rgbkrk** [comment]
In case it helps (perhaps for simplification), this has to support the VDOM spec:

* [Human Readable spec](https://github.com/nteract/vdom/blob/0.5/docs/spec.md)
* [JSON Schema](https://github.com/nteract/vdom/blob/master/vdom/schemas/vdom_schema_v1.json)

The `master` branch has some additional work that @gnestor is going to wrap up before our next release (around event binding).

File: packages/transform-vdom/src/object-to-react.ts (hunk 3 above)
Comment (rgbkrk) [reply packages/transform-vdom/src/object-to-react.ts]: For better or worse, VDOM is close to but not quite a total match to the React types. We support a subset React does, though we could allow it to expand a little bit.

**Comment by captainsafia** [comment]
OK. I think I've found a good blend for typing by leveraging some of React's built-in types.

File: packages/transform-vdom/src/object-to-react.ts [hunk 4]
```diff
@@ -35,16 +33,17 @@
...
-type ReactArray = Array<React$Element<any> | ReactArray | string>;
+interface Attributes {
+  style: object;
+  key: number;
+}
 
-type VDOMNode = VDOMEl | string | Array<VDOMNode>;
+export interface VDOMEl {
+  tagName: string; // Could be an enum honestly
+  children: React.ReactNode | VDOMEl | Array<React.ReactNode | VDOMEl>;
```
Comment (captainsafia) [inline packages/transform-vdom/src/object-to-react.ts:43]: `React.ReactNode` captures the `string` part of this (and related DOM elements). I got around the circular definition error by just copy pasting the `React.ReactNode | VDOMEl` as the value in the array *shrug*.

File: packages/transform-vdom/src/object-to-react.ts [hunk 5]
```diff
@@ -95,11 +94,11 @@ export function objectToReactElement(obj: VDOMEl): React$Element<any> {
       if (args[1] === undefined) {
         args[1] = null;
       }
-      args = args.concat(arrayToReactChildren(children));
+      args = args.concat(arrayToReactChildren(children as VDOMEl[]) as any);
```
Comment (captainsafia) [inline packages/transform-vdom/src/object-to-react.ts:97]: I had to add some typecasting here to specify which specific type within the union type we were in based on the if-statement.

File: packages/transform-vdom/src/index.js.flow [hunk 6]
```diff
@@ -0,0 +1,8 @@
+// @flow
+/* eslint-disable */
+
```
Comment (rgbkrk) [inline packages/transform-vdom/src/index.js.flow:3]: I feel like the prop types can go here too, not sure how the hand made flow defs for react components should work.

File: packages/transform-vdom/src/object-to-react.ts [hunk 7]
```diff
@@ -35,16 +33,17 @@
...
-  tagName: string, // Could be an enum honestly
-  children: VDOMNode,
-  attributes: Object,
-  key: number | string | null
-};
-
-type ReactArray = Array<React$Element<any> | ReactArray | string>;
+interface Attributes {
+  style: object;
+  key: number;
```
Comment (rgbkrk) [inline packages/transform-vdom/src/object-to-react.ts:38]: Is it ok that `Attributes` can have other properties at runtime?

**Comment by rgbkrk** [comment]
@blink1073  -- we're migrating `@nteract/transform-vdom` over to typescript. Want to check this out when you get a chance?

File: packages/transform-vdom/src/object-to-react.ts (hunk 7 above)
Comment (blink1073) [reply packages/transform-vdom/src/object-to-react.ts:38]: Should be okay, this is the contract for the minimal required properties.

**Review by blink1073 — COMMENTED** [review]
Looks reasonable to me, thanks @captainsafia and @rgbkrk!  TypeScript and React is still a somewhat shaky union, cf https://github.com/DefinitelyTyped/DefinitelyTyped/issues/28015

**Comment by rgbkrk** [comment]
> TypeScript and React is still a somewhat shaky union

We can go without `prop-types` though right? We're just using the templated props.

**Comment by blink1073** [comment]
Yup!  I was just giving you a heads up about `@types/react` being a moving target that doesn't follow semver...

**Comment by rgbkrk** [comment]
😬

**Comment by captainsafia** [comment]
@blink1073 Noted. Do you happen to know any workarounds for writing recursive references in union types in TypeScript? I managed to get a workaround for a recursive reference in an Array type but not a union type.

**Comment by blink1073** [comment]
I've never come across a recursive reference of either type.  You jumped right into the deep end :smile:   @gnestor, have you come across them?

**Comment by gnestor** [comment]
@captainsafia I have run into something like that before. 

Here's an example in phosphor: https://github.com/phosphorjs/phosphor/blob/master/packages/coreutils/src/json.ts#L10-L35

And here's the same exact example in a typescript issue: https://github.com/Microsoft/TypeScript/issues/3496#issuecomment-128553540

**Pull request merged**